File: c++/655. Print Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<string>> printTree(TreeNode* root) {
        int m = 0;
        int n = 0;
        n = dfs(root, 1, m);
        vector<vector<string>> res(m, vector<string>(n, ""));
        fill(root, 0, n-1, 0, res);
        return res;
    }

    void fill(TreeNode* root, int l , int r, int d, vector<vector<string>>& res) {
        int mid = l + (r-l)/2;
        res[d][mid] = to_string(root->val);
        if (root->left)
            fill(root->left, l, mid-1, d+1, res);
        if (root->right)
            fill(root->right, mid+1, r, d+1, res);
    }

    int dfs(TreeNode* root, int cur_depth, int &depth) {
        if (!root) return 0;
        if (!root->left && !root->right) {
            depth = max(cur_depth, depth);
            return 1;
        }
            
        int l1 = dfs(root->left, cur_depth+1, depth);
        int l2 = dfs(root->right, cur_depth+1, depth);
        return 1 + max(l1, l2)*2;
    }
};

// 1. dfs确定高度和宽度
// 2. dfs 二分填充

File: go/src/leetcode/655/655. Print Binary Tree.go
/**
@author ZhengHao Lou
Date    2022/08/22
*/
package main

import "fmt"

/**
https://leetcode.cn/problems/print-binary-tree/
*/
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func printTree(root *TreeNode) [][]string {
	var height int
	que := []*TreeNode{root}

	for ; len(que) != 0; height++ {
		for tmp := len(que); tmp != 0; tmp-- {
			node := que[0]
			que = que[1:]
			if node.Left != nil {
				que = append(que, node.Left)
			}
			if node.Right != nil {
				que = append(que, node.Right)
			}
		}
	}

	height--

	var m, n = height + 1, 1<<(height+1) - 1
	var grid = make([][]string, m)
	for i := range grid {
		grid[i] = make([]string, n)
	}

	counter := map[*TreeNode][]int{
		root: []int{0, (n - 1) >> 1},
	}

	que = []*TreeNode{root}
	for len(que) != 0 {
		for tmp := len(que); tmp != 0; tmp-- {
			node := que[0]
			que = que[1:]
			r, c := counter[node][0], counter[node][1]
			grid[r][c] = fmt.Sprintf("%v", node.Val)

			if node.Left != nil {
				counter[node.Left] = []int{r + 1, c - 1<<(height-r-1)}
				que = append(que, node.Left)
			}
			if node.Right != nil {
				counter[node.Right] = []int{r + 1, c + 1<<(height-r-1)}
				que = append(que, node.Right)
			}
		}
	}

	return grid
}

func main() {
	printTree(&TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val:   2,
			Right: &TreeNode{Val: 4},
		},
		Right: &TreeNode{
			Val: 3,
		},
	})

}
